### PR TITLE
tokenization: no double BOS tokens

### PR DESCRIPTION
**Pull request by JohannesGaessler**
Relevant discussion: https://github.com/ggerganov/llama.cpp/issues/7062

The llama.cpp tokenizer currently adds a BOS token unconditionally. However, I think it would make more sense not to do this if the prompt already starts with a BOS token since this can presumably lead to bad results. This PR makes it so that a BOS token is only added if the prompt does not already start with one.

### CODE REVIEW

**Comment by slaren** [comment]
This is not so clear to me. There is already a parameter `add_special` to configure this behavior, is the user asks for bos to be added, is it really a good idea to ignore that in some cases?

**Comment by JohannesGaessler** [comment]
Sorry, I didn't see the argument `add_special` of `llama_tokenize`. When we define a user as someone using the API specified by `llama.h` then I definitely agree that this is the wrong way to fix it. I was only looking at the perspective from using binaries like `main`, `tokenize`, or `server`. If you copy-paste an instruct format with `<|begin_of_text|>` into those you get two BOS tokens. So I think it would make more sense to fix it in those binaries?

**Comment by JohannesGaessler** [comment]
I added a function `llama_fix_double_bos` to `common.h`. The functions `llama_tokenize` in `common.h` have gained new arguments `fix_double_bos` which default to `false` and apply the double BOS fix if set to `true`. There is no change to the API in `llama.h`. I went through the files in which `llama_tokenize` is used and tried to figure out when to apply the fix. Due to the order of arguments I also had to make explicit choices for `parse_special` in some cases. Specifically:

* `common/train.cpp`, `examples/batched.swift/Sources/main.swift`: use `llama.h` API directly -> no change.
* `batched`, `beam-search`, `passkey`, `save-load-state`, `simple`: BOS was being added for the prompt, the prompt was **not** being parsed for special tokens due to the default being `false`. Now special tokens in the prompt are also being parsed. Double BOS fix added.
* `embedding`: BOS was being added for the prompt, the prompt was **not** being parsed for special tokens due to explicitly setting this to `false`. I did not change this but I think this should be set to `true` in order to be consistent with e.g. `main`, feedback would be appreciated. Double BOS fix added.
* `eval_callback`, `gritlm`, `llama-android`, `retrieval`: I don't feel confident regarding what the code actually does so I made no changes.
* `imatrix`: BOS was being added for the prompt, the prompt was **not** being parsed for special tokens due to the default being `false`. I made this to now parse special tokens; I think there are cases where either setting could cause problems (special tokens in general text corpus, training an imatrix for an instruct model on e.g. a chat log with special tokens) but I think that the latter case is more likely to happen and cause problems. Double BOS fix applied.
* `infill`, `parallel`: Special tokens were not being parsed due to the default being `false`. I enabled special token parsing and applied the double BOS fix for those tokenization calls where a BOS token is being added.
* `llava-cli`, `lookahead`, `lookup`, `lookup-create`, `lookup-stats`, `main`, `server`, `speculative`, `tokenize`: Special tokens were already being parsed, apply double BOS fix to those calls that add a BOS token.
* `perplexity`: Special tokens were not being parsed due to the default being `false`. I enabled special token parsing and applied the double BOS fix for those tokenization calls where a BOS token is being added. I think it's better to enable special token parsing since if I wanted to evaluate the effects of quantization for an instruct model I think the best way to do so would be to evaluate KL divergence over an instruct chat log in which case special tokens definitely should be parsed. I did not find instances of special tokens in Wikitext. Double BOS fix applied.
* Tests: no changes.

**Comment by ggerganov** [comment]
I'm not convinced that the `fix_double_bos` change is necessary. Typically, in the examples we don't specify the BOS token explicitly in the prompt - I believe this has been the case for a while now. Seems some of the confusion in #7062 came from people explicitly adding the BOS token in the prompt `-p`. This is more of an incorrect usage of `llama.cpp` examples and wrong expectations, so I'm not sure we should add this extra handling - the solution is to not add the BOS token explicitly

Regarding `parse_special`:

The way I think about this option is that it should be `false` when we tokenize user input (e.g. chat message, code for code completion, code for fim completion, text for perplexity calculation, etc.). It should be `true` only for the portions that are used to guide the model into some structured format - like a chat template (e.g. `<|im_start|>user`, `<|im_end|>assistant`). So technically, special tokens should not come from user inputs and even if they do, we must parse them as text - otherwise the structured format can be altered and break (i.e. prompt injection). This has been logic for setting `parse_special` the way it is on `master`

This all make sense when all special tokens such as `<|im_start|>`, `<|end_of_text|>`, `<|eot|>`, etc. are correctly defined as being `special` in the tokenizer config, but it seems there are already models out there in which the specialness of the tokens is pretty random. In such cases, our logic would break. But I think this is a problem of the model creators and again, I'm not sure we should do anything on `llama.cpp` side

Regarding `perplexity`, I don't see why we have to parse special tokens in the input text. We run `perplexity` only on base models and accidentally encountering a string that looks like a special token (for example `</s>`) in wikitext should be considered as the plain text `</s>` and not the EOS token.

What do you think?

**Comment by JohannesGaessler** [comment]
>What do you think?

I fully admit that I am not very knowledgeable about tokenization and the handling of special tokens so I will simply accept your judgement regarding what's best. This PR was fairly low-effort so feel free to close it.

Generally speaking, I would prefer to minimize the potential for user error, if only to reduce the number of bug reports about vague issues with inference results. In this particular case I invested time into an investigation where the conclusion will (unless I'm misunderstanding you) be that there were no bugs all along and that it's an issue with incorrect user assumptions about special tokens. This is time that I could have used more productively; I knew that llama.cpp adds BOS tokens but I still did not spot the double BOS tokens until after the fact. I think the most important factor for the success of an open-source project is the amount and quality of work put in by contributors so I want to minimize unproductive work.

>The way I think about this option is that it should be false when we tokenize user input (e.g. chat message, code for code completion, code for fim completion, text for perplexity calculation, etc.).

Thank you for the clarification, I think the way it's used on master does not consistently follow this principle though. I definitely did not use it that way myself when I created the files related to n-gram lookup.

>Regarding perplexity, I don't see why we have to parse special tokens in the input text. We run perplexity only on base models and accidentally encountering a string that looks like a special token (for example </s>) in wikitext should be considered as the plain text </s> and not the EOS token.

Calculating perplexity of an instruct model does not make sense I think. However, I think calculating the KL divergence of a quantized instruct model vs. the FP16 logits on an instruct dataset can potentially yield useful information; I was thinking about it in that aspect. But I forgot about `</s>`, that particular special token is short and unspecific enough that I would definitely not want to enable special token parsing for `perplexity`.

**Comment by ggerganov** [comment]
> This is time that I could have used more productively; I knew that llama.cpp adds BOS tokens but I still did not spot the double BOS tokens until after the fact. I think the most important factor for the success of an open-source project is the amount and quality of work put in by contributors so I want to minimize unproductive work.

Yeah, I see that the discussion was quite long - thanks for taking the time to look into it. 

It's probably best to invest time in improving the issue reporting process. We can provide means to generate some detailed report of a model, for example in HTML / Markdown format (see https://github.com/ggerganov/llama.cpp/issues/6715). Such a report can contain sample usage commands with the various `llama.cpp` tools, tokenization results, statistics, etc. Having such information should help both the users and the contributors a lot

> Thank you for the clarification, I think the way it's used on master does not consistently follow this principle though. I definitely did not use it that way myself when I created the files related to n-gram lookup.

`server` is also not following it exactly, but there are TODO's to fix this eventually:

https://github.com/ggerganov/llama.cpp/blob/947d3ad27d94f1addef76b5d64c314618f063933/examples/server/server.cpp#L762-L767

Maybe for `imatrix` it always makes sense to enable though because most of the datasets used for instruct models probably follow the instruction format of the models.

**Comment by JohannesGaessler** [comment]
>It's probably best to invest time in improving the issue reporting process.

Speaking of which, one of the things that I find annoying is that issues along the lines of "results wrong/bad/gibberish" frequently get people saying they have the same issue when it's really a different issue. I'm not 100% sure about how to fix this though.

**Comment by Jeximo** [comment]
> people saying they have the same issue when it's really a different issue

`llama.cpp` sees issues *even for downstream projects* - it's **a lot** to sift through.

Probably users don't recognize that the model, or how it's made may be the problem. Graphical GGUF viewer #6715 will help everyone better identify what to focus on.

**Comment by bartowski1182** [comment]
on the subject of reporting issues, it would be nice if you could somehow have a JSON structured input that would run a test on llama.cpp directly, as you said a LOT of issues are from downstream work, I've had the odd case where I almost went to report an issue but then built llama.cpp and ran ./main only to not see the issue at all

if there was a way to specify a model from HF, and a prompt, and have that run somewhere and output the results into the issue report, that could speed up a lot of confusion

**Comment by andysalerno** [comment]
Adding my two cents as a user-

In `server`, as far as I can tell, there's no option to log the final input text, post-chat-formatting. When using the OpenAI-style API, this means it's hard to be absolutely sure the chat formatting is applied correctly. So it would be great if there was some option to log the post-formatting, but pre-tokenization inputs, so we can see the final text that is being fed to the model and identify any formatting mistakes.

It might also be nice to emit warnings for the double BOS tokens scenario. It's reasonable, I think, to not fix extra BOS tokens, on the principle of garbage-in-garbage-out :D But logging a warning would at least give visibility to the issue. Same for other common formatting mistakes.

Not sure how llama.cpp is handling the Jinja templating, but if the Jinja engine supports throwing errors, it could be nice to encourage model makers to write chat templates that throw errors if validations fail (like how Mistral's template will throw if the input is not perfectly alternating user/assistant/user/assistant). Having it blow up with an obvious error instead of silently continuing would catch mistakes a lot faster.

**Comment by reuank** [comment]
>  So it would be great if there was some option to log the post-formatting, but pre-tokenization inputs, so we can see the final text that is being fed to the model and identify any formatting mistakes.

I 100% agree with @andysalerno on this. In a project, I am applying a chat template myself before giving the prompt to llama.cpp. I have seen a massive drop in output quality (in this case: the performance on a multiple choice benchmark with a chain of thought step in between) when using the template compared to when I don't use it, but it's hard to find out what is going wrong. It would be great to add an option that echoes the exact prompt (and its token ids) that was fed to the model.

I think https://github.com/ggerganov/llama.cpp/pull/7332 is a good first step, but having more debug options would be even more helpful.

**Comment by JohannesGaessler** [comment]
For binaries like `main` there already is the `--verbose-prompt` option but I'm not seeing it in the server documentation (but someone feel free to correct me).

**Comment by abc-nix** [comment]
> For binaries like `main` there already is the `--verbose-prompt` option but I'm not seeing it in the server documentation (but someone feel free to correct me).

With the `--verbose` option server does log the input.

```
{"tid":"140469774090240","timestamp":1715890225,"level":"VERB","function":"update_slots","line":1948,"msg":"prompt tokenized","id_slot":0,"id_task":0,"n_ctx":8192,"n_keep":0,"n_prompt_tokens":52,"prompt_tokens":"<|begin_of_text|>This is a conversation between User and Llama, a friendly chatbot. Llama is helpful, kind, honest, good at writing, and never fails to answer any requests immediately and with precision.\n\nUser: Write only 3 words.\nLlama:"}
```

**Comment by reuank** [comment]
@abc-nix Thanks for the hint! I added the `--verbose` server option to the README: https://github.com/ggerganov/llama.cpp/pull/7335

However, I would still prefer an option to get the processed and tokenized prompt in the server response JSON.
